### ISSUE TITLE
Feature request: Empty line setting for minimum required properties

### ISSUE THREAD

**Issue by justrhysism** [hudochenkov/stylelint-order#86]
I like separating the properties into rational groups with lines in between - however when there's only a handful (< 3 || 4) of properties the gaps are a bit unwieldily and it makes it difficult to reason.

Suggestion: a minimum property setting for enacting the empty lines.

e.g. a threshold setting of 3 (where `ddd` and `eee` are their own group):

```scss
// Only 3 properties, so do not insert lines
a {
  aaa: 0;
  bbb: 0;
  eee: 0;
}

// Greater than 3 properties, so insert new lines
a {
  aaa: 0;
  bbb: 0;
  ccc: 0;

  ddd: 0;
  eee: 0;
}
```
---

Original request on *stylelint-config-rational-order*: https://github.com/constverum/stylelint-config-rational-order/issues/11

**Comment by justrhysism** [comment]
So I've created a draft PR with a working prototype to facilitate this feature. Tests are located in `~/rules/properties-order/tests/empty-line-minimum-property-threshold.js`

Currently there are two failing test cases which I'm not too sure which direction should be taken:

### Lines above comments (9)

```css
a {
  /* comment */
  display: none;

  /* comment */
  position: absolute;
}
```

### Lines above variables (13.4)

```scss
a {
  position: absolute;

  $display: none;
}
```

Before I go to the effort of figuring out the logic to implement this (the level of effort so far has been pretty minimal), I'd like to **ask if these are cases in which the empty lines should be retained**.

_I've kept the discussion on functionality on this issue to leave the PR discussion for code quality suggestions/changes._

cc/ @jm3, @hudochenkov

**Comment by hudochenkov** [comment]
It's an interesting problem to solve.

It has many different cases to support. As you can see in the following problems there is a lot of things needed to be solved.

Dollar variables and custom properties ignored by this rule.

## Problem 1

Imagine following CSS groupped correctly and empty lines as it should be:

```css
a {
	aaa: 0;
	bbb: 0;

	ccc: 0;
	
	ddd: 0;
	
	eee: 0;
	
	fff: 0;
	
	ggg: 0;
	hhh: 0;
}
```

What plugin will produce with option set to 2?

Glue together all lone properties:

```css
a {
	aaa: 0;
	bbb: 0;

	ccc: 0;
	ddd: 0;
	eee: 0;
	fff: 0;
	
	ggg: 0;
	hhh: 0;
}
```

Glue together lone properties until they meet option value:

```css
a {
	aaa: 0;
	bbb: 0;

	ccc: 0;
	ddd: 0;

	eee: 0;
	fff: 0;
	
	ggg: 0;
	hhh: 0;
}
```

Glue lone properties to a previous group which meet option value:

```css
a {
	aaa: 0;
	bbb: 0;
	ccc: 0;
	ddd: 0;
	eee: 0;
	fff: 0;
	
	ggg: 0;
	hhh: 0;
}
```

## Problem 2

Imagine following CSS groupped correctly and empty lines as it should be:

```css
a {
	aaa: 0;
	bbb: 0;

	ccc: 0;
	
	ddd: 0;
	
	eee: 0;
	
	fff: 0;
	
	ggg: 0;
	hhh: 0;
}
```

What plugin will produce with option set to 3?

Glue everything together:


```css
a {
	aaa: 0;
	bbb: 0;
	ccc: 0;
	ddd: 0;
	eee: 0;
	fff: 0;
	ggg: 0;
	hhh: 0;
}
```

Glue together lone properties until they meet option value:

```css
a {
	aaa: 0;
	bbb: 0;
	ccc: 0;
	
	ddd: 0;
	eee: 0;
	fff: 0;
	
	ggg: 0;
	hhh: 0;
}
```

## Problem 3

Imagine following CSS groupped correctly and empty lines as it should be:

```css
a {
	aaa: 0;
	bbb: 0;

	/* comment */
	ccc: 0;

	/* comment */
	ddd: 0;
	
	eee: 0;

	fff: 0;
	ggg: 0;
}
```

What plugin will produce with option set to 2?

No changes, because properties with comments are ok to be alone.

```css
a {
	aaa: 0;
	bbb: 0;

	/* comment */
	ccc: 0;

	/* comment */
	ddd: 0;

	eee: 0;

	fff: 0;
	ggg: 0;
}
```

Glued together, ignoring comments:

```css
a {
	aaa: 0;
	bbb: 0;

	/* comment */
	ccc: 0;
	/* comment */
	ddd: 0;
	eee: 0;

	fff: 0;
	ggg: 0;
}
```

Glue only property without a comment:

```css
a {
	aaa: 0;
	bbb: 0;

	/* comment */
	ccc: 0;

	/* comment */
	ddd: 0;
	eee: 0;

	fff: 0;
	ggg: 0;
}
```

**Comment by justrhysism** [comment]
@hudochenkov I think you've misunderstood the suggestion; I probably should've provided examples in my original post.

I don't want to set a minimum size of the _groups_, but rather a minimum number (threshold) of properties in the selector _required_ for the grouping to be invoked.

e.g. a threshold setting of 3 (where `ddd` and `eee` are their own group):

```scss
// Only 3 properties, so do not insert lines
a {
  aaa: 0;
  bbb: 0;
  eee: 0;
}

// Greater than 3 properties, so insert new lines
a {
  aaa: 0;
  bbb: 0;
  ccc: 0;

  ddd: 0;
  eee: 0;
}
```

**Comment by hudochenkov** [comment]
Now I get it :) There are question to think about.

If we have only 3 property in the rule, when option is set to 4. Does it mean we'll remove empty lines? Does it mean new this option is targeted only for people, who has groups in config and each group has `emptyLineBefore: "always"`?

Let's assue that the following input is correct, and it's three groups with empty line before each. Threshold is disabled.

```css
a {
	aaa: 0;

	/* comment */
	bbb: 0;

	/* comment */
	ccc: 0;
}
```

With threshold set to 4 do we remove empty lines?

---

Threshold is 4. We have only 3 property inside.

```css
a {
	aaa: 0;

	--num: 0;

	bbb: var(--num);

	ccc: var(--num);
}
```

What is the output?

Unchanged:

```css
a {
	aaa: 0;

	--num: 0;

	bbb: var(--num);

	ccc: var(--num);
}
```

Remove before `ccc`:

```css
a {
	aaa: 0;

	--num: 0;

	bbb: var(--num);
	ccc: var(--num);
}
```

Remove before `bbb` and `ccc`:

```css
a {
	aaa: 0;

	--num: 0;
	bbb: var(--num);
	ccc: var(--num);
}
```

Remove all empty lines:

```css
a {
	aaa: 0;
	--num: 0;
	bbb: var(--num);
	ccc: var(--num);
}
```

Why?

**Comment by justrhysism** [comment]
@hudochenkov exactly why this needs thought and discussion 👍 
 
When starting this journey, I hadn't realised that _individual groups_ were configured with `emptyLineBefore: "always"`. Maybe we could add an option for `emptyLineBefore: "threshold"`?

To preface: for the sake of discussion I'm going to ignore technical considerations for now, as there are many cases to discuss.
---

### Case; threshold: 4; properties: 3; with comments

```css
a {
	aaa: 0;

	/* comment */
	bbb: 0;

	/* comment */
	ccc: 0;
}
```
**Thought:** I think in the case of _comments_ the lines should **not be removed**.

**Reasoning:** typically such comments are there for a reason, and removing the lines here would reduce readability.

---

### Case; threshold: 4; 3 properties; with custom property

```css
a {
	aaa: 0;

	--num: 0;

	bbb: var(--num);

	ccc: var(--num);
}
```

I haven't got an answer for this one just now. I guess it depends on how you interpret the custom property. Custom properties are indeed _properties_, but mentally, they are really _variables_.

I think I need to get a better understanding of how custom properties are treated currently before taking a position. One _could_ argue that they are not treated any differently to other properties?

@hudochenkov I'd like to know your thoughts if you have any.

Thanks again for a great discussion.

**Comment by hudochenkov** [comment]
Custom properties (or any other nodes, except standard properties) should not be affected by `properties-order`.

This feature will have to many cases to worry about. I think it will be hard to make it right and robust.

P. S. To be honest, I'm not interested in this idea, because I'm not going to use it myself. And I don't want to explore solutions as well. But if you solve the problem we can include it in the plugin.

**Comment by justrhysism** [comment]
> And I don't want to explore solutions as well. 

Well that's disappointing.

> But if you solve the problem we can include it in the plugin.

@hudochenkov I'm happy to solve the problems, but I need to know what your expectations are for acceptance; hence the need for discussion. 

Already through our discussion you've raised things I hadn't thought about. As the author of the plugin, you have the knowledge depth for the cases which the rest of us simply won't think about.

**Comment by justrhysism** [comment]
Okay, so it turned out to be pretty simple. The threshold option adds/removes lines in the same manner as the `always` and `never` settings, by toggling their respective logic if:

- The group is set to `emptyLineBefore: "threshold"`
- `emptyLineMinimumPropertyThreshold > 0`
- The number of properties is:
  - above threshold: triggers `always`
  - below threshold: triggers `never`

The behaviour is also applied to the unspecified setting.

Given this feature is entirely opt-in, it does not affect existing installations.

**Comment by justrhysism** [comment]
Closed #88

**Comment by hudochenkov** [comment]
`3.1.0` released.
